ado/ipatrackversions.ado

```diff
@@ -0,0 +1,195 @@
+	survey */
+	
+	
+
+//define inputs
+version 13
+
+#delimit ;
+syntax varname,  //varname is the form version variable - must be num. 
+	/* specify uid for the survey, enumerator id */
+    id(varname) ENUMerator(varname) 
+	/* list of other vars to keep */
+	[KEEPvars(string)] 
+	/* specify date var, i.e. submission date var */
+	submit(varname numeric)
+	/* output filename */
+	saving(string) 
+	;	
+#delimit cr
+
+	
+	di ""
+	di "Compiling information on survey form versions by submission date..."
+
+	qui {
+	
+	* test for fatal conditions 
+	cap confirm numeric var `varlist'
+	if _rc {
+		di as err "Variable used for form version (`varlist') not numeric. Must be numeric and ordinal for the program to work correctly"
+		error 101
+	}
+		
+	if !mi("`keep'") {
+		foreach var of varlist `keep' {
+			cap confirm var `var'
+			if _rc {
+				di as err `"Tried to specify keeping the var "`var'" - "`var'" does not exist in the dataset"'
+				error 101
+			}
+		}
+	}
+	
+	count if `submit' == . 
+	if `r(N)' > 0 {
+		di as err `"There are missing values of `submit'. Either drop these observations or restrict them using an "if" statement."'
+		error 101
+	}
+
+	
+	* initialize tempvars  
+	tempvar header_submit header_fvs formatted_subdate ///
+	outdated_fvs_header max_fv_by_subdate wrong_fv wrong_fv_today 
+	
+	* export sheet headers 
+	gen `header_submit' = . 
+	gen `header_fvs' = . 
+	lab var `header_submit' "Submission Date" 
+	lab var `header_fvs' "Form Versions" 	
+	export excel `header_submit' `header_fvs'  using "`saving'" in 1, sheet("T3. form versions") firstrow(varl) sheetreplace	
+
+	* convert `header_submit' to %td format if needed	
+	ds `submit', has(format %td*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = `submit'
+		}
+
+	ds `submit', has(format %tc*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofc(`submit')
+		}
+		
+	ds `submit', has(format %tC*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofC(`submit')
+		}
+		
+	ds `submit', has(format %tb*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofb(`submit')
+		}
+	
+	ds `submit', has(format %tw*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofw(`submit')
+		}
+		
+	ds `submit', has(format %tm*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofm(`submit')
+		}
+
+	ds `submit', has(format %tq*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofq(`submit')
+		}
+		
+	ds `submit', has(format %th*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofh(`submit')
+		}
+
+	ds `submit', has(format %ty*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofy(`submit')
+		}
+	
+	format `formatted_subdate' %tdCCYY/NN/DD
+	
+	tab `formatted_subdate' `varlist'
+	if mi("`r(N)'") {
+		di as err `"No observations in cross-tab of `submit' and `varlist' - check your data"'
+		error 122
+	}
+	
+	* format and export submission dates (left hand column of table)
+	preserve 
```

Comment:
I think all the `preserve`  and `restore` statements could be replaced by using a single temporary file...